src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteFactory.cs

```diff
@@ -350,6 +355,9 @@ private static CallSiteResultCacheLocation GetCommonCacheLocation(CallSiteResult
             Justification = "MakeGenericType here is used to create a closed generic implementation type given the closed service type. " +
             "Trimming annotations on the generic types are verified when 'Microsoft.Extensions.DependencyInjection.VerifyOpenGenericServiceTrimmability' is set, which is set by default when PublishTrimmed=true. " +
             "That check informs developers when these generic types don't have compatible trimming annotations.")]
+        [UnconditionalSuppressMessage("AotAnalysis", "IL3050:RequiresDynamicCode",
+            Justification = "When 'Microsoft.Extensions.DependencyInjection.VerifyAotCompatibility' is set, which is set by default when PublishAot=true, " +
+            "this method ensures the generic types being created aren't using ValueTypes.")]
```

Comment:
Are we backing off from build check and moving back to runtime exceptions for AOT in general or is this somehow a unique case that nobody will mimic?

Reply:
> Are we backing off from build check

No, definitely not. The build/publish time warnings are still the main tool we have for telling app devs what will and won't work with NativeAOT.

> is this somehow a unique case that nobody will mimic?

This probably won't be the only case where we take this approach, but I expect these situations to be extremely rare. (Note, as called out in #79286, we took this same approach to resolve trimming warnings in this library as well.)

The situation here is that:

1. DependencyInjection is a very popular library - all of ASP.NET is built around it.
2. DependencyInjection has 2 rarely used, edge-case features that aren't compatible with NativeAOT.
3. As it exists today, and what we did for .NET 7, the whole library/feature needs to be marked as `RequiresDynamicCode` because of these 2 rarely used features.

That isn't an acceptable solution going forward because the library works for the main-line scenarios and we want to enable ASP.NET applications to work with NativeAOT without warnings. Always warning from this library would "numb" developers to AOT warnings, since their app works even though it warns.

Reply:
I would refer to this as the "pragmatic approach".

Reply:
> I would refer to this as the "pragmatic approach".

That's fair, but should not we use this approach for anyone using MakeGenericType? It seems like the current approach of triggering cascading warning everywhere is too much when it can actually only happen for value-type arguments.

Reply:
I think waiting until runtime for all MakeGenericType/MakeGenericMethod errors in the hope that it won't get used with value types would probably be going too far, because those methods are commonly used with value types.  [For example](https://github.com/dotnet/aspnetcore/blob/63d2fb07e93007af1f016d447a904ce187256881/src/Http/Http.Extensions/src/RequestDelegateFactory.cs#L611-L626), ASP.NET Cores calls MakeGenericMethod with route handlers' parameter types which could easily be value types when IsDynamicCodeSupported, and that's just one of many similar usages.

This approach does have the downside of not even firing at runtime in development/testing if the right/wrong code paths don't get hit. But in the DI case, I think open generic services with value type arguments are exceedingly rare relative to the number of people using DI, so it's worth accepting this downside to avoid all the false positives.

Reply:
This code isn't AOT safe though. I think in .NET 8, once we have the source generator, we'd get rid of these checks in the codebase.